File: LeetCodePractice/src/dfs/MakingaLargeIsland.java
package dfs;

import java.util.*;

/*
In a 2D grid of 0s and 1s, we change at most one 0 to a 1.

After, what is the size of the largest island? (An island is a 4-directionally connected group of 1s).

Example 1:

Input: [[1, 0], [0, 1]]
Output: 3
Explanation: Change one 0 to 1 and connect two 1s, then we get an island with area = 3.
Example 2:

Input: [[1, 1], [1, 0]]
Output: 4
Explanation: Change the 0 to 1 and make the island bigger, only one island with area = 4.
Example 3:

Input: [[1, 1], [1, 1]]
Output: 4
Explanation: Can't change any 0 to 1, only one island with area = 4.
 

Notes:

1 <= grid.length = grid[0].length <= 50.
0 <= grid[i][j] <= 1.

analysis:
mark different island with different color
for each 0 find its island neighbors and connect them all
*/
public class MakingaLargeIsland {
	int rows, cols;
	Map<Integer, Integer> area;
	int[] dirs = { 1, 0, -1, 0, 1 };

	public int largestIsland(int[][] grid) {
		rows = grid.length;
		cols = grid[0].length;
		area = new HashMap<>();
		int mark = 1;
		int res = 1;
		for (int r = 0; r < rows; r++) {
			for (int c = 0; c < cols; c++) {
				if (grid[r][c] == 1) {
					mark++;
					area.put(mark, getArea(grid, mark, r, c));
					res = Math.max(res, area.get(mark));
				}
			}
		}
		for (int r = 0; r < rows; r++) {
			for (int c = 0; c < cols; c++) {
				if (grid[r][c] == 0) {
					int total = 1;
					Set<Integer> nbColorSet = new HashSet<>();
					for (int i = 0; i < dirs.length - 1; i++) {
						nbColorSet.add(getColor(grid, r + dirs[i], c + dirs[i + 1]));
					}
					for (int color : nbColorSet) {
						total += area.getOrDefault(color, 0);
					}
					res = Math.max(res, total);
				}
			}
		}
		return res;
	}

	private int getColor(int[][] grid, int r, int c) {
		if (r >= rows || r < 0 || c >= cols || c < 0) {
			return 0;
		} else {
			return grid[r][c];
		}
	}

	private int getArea(int[][] grid, int color, int r, int c) {
		if (r >= rows || r < 0 || c >= cols || c < 0 || grid[r][c] != 1) {
			return 0;
		}
		grid[r][c] = color;
		int res = 1;
		for (int i = 0; i < dirs.length - 1; i++) {
			res += getArea(grid, color, r + dirs[i], c + dirs[i + 1]);
		}
		return res;
	}

	//////////////////slower way ///////////////////////////////////////////////
	public int largestIslandII(int[][] grid) {
		int res = -1;
		for (int r = 0; r < rows; r++) {
			for (int c = 0; c < cols; c++) {
				if (grid[r][c] == 0) {
					grid[r][c] = 1;
					res = Math.max(res, dfs(grid, r, c, new boolean[rows][cols]));
					grid[r][c] = 0;
				}
			}
		}
		return res == -1 ? rows * cols : res;
	}

	private int dfs(int[][] grid, int r, int c, boolean[][] visited) {
		if (r >= rows || r < 0 || c >= cols || c < 0 || visited[r][c] || grid[r][c] == 0) {
			return 0;
		}
		visited[r][c] = true;
		int res = 1;
		for (int i = 0; i < dirs.length - 1; i++) {
			res += dfs(grid, r + dirs[i], c + dirs[i + 1], visited);
		}
		return res;
	}
}

File: LeetCodePython/MakingaLargeIsland.py
"""
You are given an n x n binary matrix grid. You are allowed to change at most one 0 to be 1.

Return the size of the largest island in grid after applying this operation.

An island is a 4-directionally connected group of 1s.



Example 1:

Input: grid = [[1,0],[0,1]]
Output: 3
Explanation: Change one 0 to 1 and connect two 1s, then we get an island with area = 3.
Example 2:

Input: grid = [[1,1],[1,0]]
Output: 4
Explanation: Change the 0 to 1 and make the island bigger, only one island with area = 4.
Example 3:

Input: grid = [[1,1],[1,1]]
Output: 4
Explanation: Can't change any 0 to 1, only one island with area = 4.


Constraints:

n == grid.length
n == grid[i].length
1 <= n <= 500
grid[i][j] is either 0 or 1.
"""
import collections
from typing import List


class MakingaLargeIsland:

    def largestIsland(self, grid: List[List[int]]) -> int:
        self.rows, self.cols = len(grid), len(grid[0])
        self.dirs = [(0, 1), (0, -1), (-1, 0), (1, 0)]
        color, res, memo = 1, 0, {0: 0}
        for r in range(0, self.rows):
            for c in range(0, self.cols):
                if grid[r][c] == 1:
                    color += 1
                    memo[color] = self.getArea(r, c, color, grid)
                    res = max(res, memo[color])

        for r in range(0, self.rows):
            for c in range(0, self.cols):
                if grid[r][c] == 0:
                    total = 1
                    nb_colors = set()
                    for dir in self.dirs:
                        nb_r, nb_c = r + dir[0], c + dir[1]
                        if 0 <= nb_r < self.rows and 0 <= nb_c < self.cols:
                            nb_colors.add(grid[nb_r][nb_c])
                    for nb_color in nb_colors:
                        total += memo[nb_color]
                    res = max(res, total)
        return res

    def getArea(self, r, c, color, grid):
        res = 1
        grid[r][c] = color
        for dir in self.dirs:
            nb_r, nb_c = r + dir[0], c + dir[1]
            if 0 <= nb_r < self.rows and 0 <= nb_c < self.cols and grid[nb_r][nb_c] == 1:
                res += self.getArea(nb_r, nb_c, color, grid)
        return res
